File: Haskell/Paciente.hs
module Paciente where

import Data.List

data Paciente = Paciente {nome:: String, sexo:: String, cpf::String, endereco:: String, idade:: Int, telefone::String} deriving(Show, Eq)

adicionaPaciente:: String -> String -> String -> String -> Int -> String -> Paciente
adicionaPaciente nome sexo cpf endereco idade telefone = 
    (Paciente {nome = nome,
    sexo = sexo,
    cpf = cpf,
    endereco = endereco,
    idade = idade,
    telefone = telefone
})

todosOsPacientes:: [Paciente] -> String
todosOsPacientes [] = ""
todosOsPacientes (h:t) = "Nome: " ++ nome h ++ ", " ++ 
                        "Sexo: " ++ show(sexo h) ++ ", " ++ 
                        "CPF: " ++ cpf h ++ ", " ++ 
                        "Endereço: " ++ endereco h ++ ", " ++ 
                        "Idade: " ++ show(idade h) ++ ", " ++ 
                        "Telefone: " ++ telefone h ++ "\n" ++ 
                        todosOsPacientes t


pegaIdadePaciente :: String -> [Paciente] -> Int
pegaIdadePaciente _ [] = 0
pegaIdadePaciente cpfPaciente (h:t)
    | cpf h == cpfPaciente = idade h
    |otherwise = pegaIdadePaciente cpfPaciente t

pegaPacientesHabilitados :: Int -> [Paciente] -> String
pegaPacientesHabilitados _ [] = ""
pegaPacientesHabilitados faixaEtaria (h:t)
                        | idade h >= faixaEtaria = "Nome: | " ++ nome h ++ " | Telefone: | " ++ telefone h ++ " | Sexo: " ++ sexo h ++ "\n" ++ pegaPacientesHabilitados faixaEtaria t ++ "\n"
                        | otherwise = pegaPacientesHabilitados faixaEtaria t ++ "\n"

atualizaEnderecoPaciente :: String -> [Paciente] -> String -> [Paciente]
atualizaEnderecoPaciente cpfPaciente [] novoEndereco = []
atualizaEnderecoPaciente cpfPaciente (h:t) novoEndereco 
    | cpfPaciente == cpf h = ((Paciente(nome h) (sexo h) (cpf h) (novoEndereco) (idade h) (telefone h)) : t)
    | otherwise =  (h: atualizaEnderecoPaciente cpfPaciente t novoEndereco)

atualizaNomePaciente :: String -> [Paciente] -> String -> [Paciente]
atualizaNomePaciente cpfPaciente [] novoValor = []
atualizaNomePaciente cpfPaciente (h:t) novoValor 
    | cpfPaciente ==  cpf h = ((Paciente(novoValor) (sexo h) (cpf h) (endereco h) (idade h) (telefone h)) : t)
    | otherwise =  (h: atualizaNomePaciente cpfPaciente t novoValor)

atualizaSexoPaciente :: String -> [Paciente] -> String -> [Paciente]
atualizaSexoPaciente cpfPaciente [] novoValor = []
atualizaSexoPaciente cpfPaciente (h:t) novoValor 
    | cpfPaciente == cpf h = ((Paciente(nome h) (novoValor) (cpf h) (endereco h) (idade h) (telefone h)) : t)
    | otherwise =  (h: atualizaSexoPaciente cpfPaciente t novoValor)

atualizaTelefonePaciente :: String -> [Paciente] -> String -> [Paciente]
atualizaTelefonePaciente cpfPaciente [] novoValor = []
atualizaTelefonePaciente cpfPaciente (h:t) novoValor 
    | cpfPaciente == cpf h = ((Paciente(nome h) (sexo h) (cpf h) (endereco h) (idade h) (novoValor)) : t)
    | otherwise =  (h: atualizaTelefonePaciente cpfPaciente t novoValor)

atualizaIdadePaciente :: String -> [Paciente] -> Int -> [Paciente]
atualizaIdadePaciente cpfPaciente [] novoValor = []
atualizaIdadePaciente cpfPaciente (h:t) novoValor 
    | cpfPaciente == cpf h = ((Paciente(nome h) (sexo h) (cpf h) (endereco h) (novoValor) (telefone h)) : t)
    | otherwise =  (h: atualizaIdadePaciente cpfPaciente t novoValor)

File: Prolog/Paciente.pl
:- style_check(-singleton).
:- style_check(-discontiguous).

constroiPaciente(Nome, Sexo, CPF, Endereco, Idade, Telefone, paciente(Nome, Sexo, CPF, Endereco, Idade, Telefone)).

/* retorna os dados do Paciente */
getPacienteNome(paciente(Nome, _, _, _, _, _), Nome).
getPacienteSexo(paciente(_, Sexo, _, _, _, _), Sexo).
getPacienteCPF(paciente(_, _, CPF, _, _, _), CPF).
getPacienteEndereco(paciente(_, _, _, Endereco, _, _), Endereco).
getPacienteIdade(paciente(_, _, _, _, Idade, _), Idade).
getPacienteTelefone(paciente(_, _, _, _, _, Telefone), Telefone).
%getIddPaciente(_,[],ListaVacinacao):- write('Paciente não existe.').
getIddPaciente(_,[],IdadePaciente) :- nl.
getIddPaciente(Cpf,[H|T],IdadePaciente):- getPacienteCPF(H,CpfPacienteH),
    (Cpf = CpfPacienteH -> getPacienteIdade(H,Idade),IdadePaciente is Idade;
    getIddPaciente(Cpf,T,IdadePaciente)).

% pega paciente
getPaciente(CpfPaciente, [], Paciente).
getPaciente(CpfPaciente, [H|T], Paciente):-
    getPacienteCPF(H,Cpf), string_upper(Cpf, CpfUpper), string_upper(CpfPaciente, CpfPacienteUpper),
    (CpfUpper == CpfPacienteUpper -> Paciente = H; getPaciente(CpfPaciente, T, Paciente)).

% busca paciente
buscaPaciente(_,[],Resultado):- Resultado = 'Paciente não encontrado.'.
buscaPaciente(Cpf,[H|T], Resultado):- getPacienteCPF(H,CpfPaciente),pacienteToString(H,PacienteToString),
            (Cpf = CpfPaciente -> Resultado is PacienteToString; buscaPaciente(NomePaciente,T,Resultado)).

% lista os pacientes
listarPacientes([]):- nl.
listarPacientes([H|T]):- pacienteToString(H,PacienteToString), write(PacienteToString), nl, listarPacientes(T).

% representação em texto do paciente
pacienteToString(paciente(Nome,Sexo, CPF, Endereco, Idade, Telefone), Resultado) :- 
    string_concat('Nome: ', Nome, Concat1),
    string_concat(Concat1, ' - Sexo: ', Concat2), 
    string_concat(Concat2, Sexo, Concat3),
    string_concat(Concat3, ' - CPF: ', Concat4), 
    string_concat(Concat4, CPF, Concat5),
    string_concat(Concat5, ' - Endereço: ', Concat6),
    string_concat(Concat6, Endereco, Concat7),
    string_concat(Concat7, ' - Idade: ', Concat8),
    string_concat(Concat8, Idade, Concat9),
    string_concat(Concat9, ' - Telefone: ', Concat10),
    string_concat(Concat10, Telefone, Resultado).

deletaPaciente([], ListaAtual, CpfPaciente, NovaLista).
deletaPaciente([H|T], ListaAtual, CpfPaciente, NovaLista):- 
    getPacienteCPF(H,Cpf), string_upper(Cpf, CpfUpper), string_upper(CpfPaciente, CpfPacienteUpper),
    (CpfUpper == CpfPacienteUpper -> delete(ListaAtual,H, NovaLista);
    deletaPaciente(T, ListaAtual, CpfPaciente, NovaLista)).

atualizaNomePaciente([], CpfPaciente, NovoValor, PacienteSaida).
atualizaNomePaciente([H|T], CpfPaciente, NovoValor, PacienteSaida):-
    getPacienteCPF(H,Cpf), string_upper(Cpf, CpfUpper), string_upper(CpfPaciente, CpfPacienteUpper),
    (CpfUpper == CpfPacienteUpper -> novoNomePaciente(H, NovoValor, PacienteSaida), pacienteToString(PacienteSaida,Result), write(Result), nl;
    atualizaNomePaciente(T, CpfPaciente, NovoValor, PacienteSaida)).

novoNomePaciente(Paciente, NovoValor, Result):- 
    getPacienteNome(Paciente, NOME),
    getPacienteCPF(Paciente, CPF),
    getPacienteEndereco(Paciente, ENDERECO),
    getPacienteIdade(Paciente, IDADE),
    getPacienteSexo(Paciente, SEXO),
    getPacienteTelefone(Paciente, TELEFONE),
    constroiPaciente(NovoValor, SEXO, CPF, ENDERECO, IDADE, TELEFONE, Result).

atualizaSexoPaciente([], CpfPaciente, NovoValor, PacienteSaida).
atualizaSexoPaciente([H|T], CpfPaciente, NovoValor, PacienteSaida):-
    getPacienteCPF(H,Cpf), string_upper(Cpf, CpfUpper), string_upper(CpfPaciente, CpfPacienteUpper),
    (CpfUpper == CpfPacienteUpper -> novoSecsoPaciente(H, NovoValor, PacienteSaida), pacienteToString(PacienteSaida,Result), write(Result), nl;
    atualizaSexoPaciente(T, CpfPaciente, NovoValor, PacienteSaida)).

novoSecsoPaciente(Paciente, NovoValor, Result):- 
    getPacienteNome(Paciente, NOME),
    getPacienteCPF(Paciente, CPF),
    getPacienteEndereco(Paciente, ENDERECO),
    getPacienteIdade(Paciente, IDADE),
    getPacienteSexo(Paciente, SEXO),
    getPacienteTelefone(Paciente, TELEFONE),
    constroiPaciente(NOME, NovoValor, CPF, ENDERECO, IDADE, TELEFONE, Result).

atualizaTelefonePaciente([], CpfPaciente, NovoValor, PacienteSaida).
atualizaTelefonePaciente([H|T], CpfPaciente, NovoValor, PacienteSaida):-
    getPacienteCPF(H,Cpf), string_upper(Cpf, CpfUpper), string_upper(CpfPaciente, CpfPacienteUpper),
    (CpfUpper == CpfPacienteUpper -> novoTelPaciente(H, NovoValor, PacienteSaida), pacienteToString(PacienteSaida,Result), write(Result), nl;
    atualizaTelefonePaciente(T, CpfPaciente, NovoValor, PacienteSaida)).

novoTelPaciente(Paciente, NovoValor, Result):- 
    getPacienteNome(Paciente, NOME),
    getPacienteCPF(Paciente, CPF),
    getPacienteEndereco(Paciente, ENDERECO),
    getPacienteIdade(Paciente, IDADE),
    getPacienteSexo(Paciente, SEXO),
    getPacienteTelefone(Paciente, TELEFONE),
    constroiPaciente(NOME, SEXO, CPF, ENDERECO, IDADE, NovoValor, Result).

atualizaEnderecoPaciente([], CpfPaciente, NovoValor, PacienteSaida).
atualizaEnderecoPaciente([H|T], CpfPaciente, NovoValor, PacienteSaida):-
    getPacienteCPF(H,Cpf), string_upper(Cpf, CpfUpper), string_upper(CpfPaciente, CpfPacienteUpper),
    (CpfUpper == CpfPacienteUpper -> novoEndPaciente(H, NovoValor, PacienteSaida), pacienteToString(PacienteSaida,Result), write(Result), nl;
    atualizaEnderecoPaciente(T, CpfPaciente, NovoValor, PacienteSaida)).

novoEndPaciente(Paciente, NovoValor, Result):- 
    getPacienteNome(Paciente, NOME),
    getPacienteCPF(Paciente, CPF),
    getPacienteEndereco(Paciente, ENDERECO),
    getPacienteIdade(Paciente, IDADE),
    getPacienteSexo(Paciente, SEXO),
    getPacienteTelefone(Paciente, TELEFONE),
    constroiPaciente(NOME, SEXO, CPF, NovoValor, IDADE, TELEFONE, Result).

atualizaIdadePaciente([], CpfPaciente, NovoValor, PacienteSaida).
atualizaIdadePaciente([H|T], CpfPaciente, NovoValor, PacienteSaida):-
    getPacienteCPF(H,Cpf), string_upper(Cpf, CpfUpper), string_upper(CpfPaciente, CpfPacienteUpper),
    (CpfUpper == CpfPacienteUpper -> novoIdadePaciente(H, NovoValor, PacienteSaida), pacienteToString(PacienteSaida,Result), write(Result), nl;
    atualizaIdadePaciente(T, CpfPaciente, NovoValor, PacienteSaida)).

novoIdadePaciente(Paciente, NovoValor, Result):- 
    getPacienteNome(Paciente, NOME),
    getPacienteCPF(Paciente, CPF),
    getPacienteEndereco(Paciente, ENDERECO),
    getPacienteIdade(Paciente, IDADE),
    getPacienteSexo(Paciente, SEXO),
    getPacienteTelefone(Paciente, TELEFONE),
    constroiPaciente(NOME, SEXO, CPF, ENDERECO, NovoValor, TELEFONE, Result).

listaPacientesASeremVacinados(_,[]):- nl.
listaPacientesASeremVacinados(IdadeMinima,[H|T]):- getPacienteIdade(H,IdadeH),
    (IdadeMinima =< IdadeH -> pacienteToString(H,PacienteToString),write(PacienteToString),nl,listaPacientesASeremVacinados(IdadeMinima,T);
    listaPacientesASeremVacinados(IdadeMinima,T)).
